crates/runtime/src/instance/allocator.rs

```diff
@@ -486,6 +506,13 @@ fn initialize_instance(
     Ok(())
 }
 
+/// Initialize the VMContext data associated with an Instance.
+///
+/// Precondition: the VMContext memory must be zeroed. We omit writes
+/// here to fields that should be initialized to zero. This allows the
+/// caller to use an efficient means of zeroing memory (such as using
+/// anonymous-mmap'd zero pages) if available, rather than falling
+/// back onto a memset (or the manual equivalent) here.
```

Comment:
Right now I think that this may be detrimental to the on-demand allocator because it either does a too-large `memset` right now for the entire `VMContext` or it will need to call `calloc`, which also does a too-large `memset`.

An alternative design, though, would be to pass a flag into this function whether the memory is known to be zero'd. That way we could conditionally zero out the array of `VMCallerCheckedAnyfunc` contents based on whether it was pre-zeroed already, making the pooling allocator efficiently use `madvise` while the on-demand allocator still initializes as little memory as possible.

After saying this, though, I think it may actually be best to take this incrementally? What do you think about using `memset` here to zero memory and benchmarking later how much faster using `madvise` is for the pooling allocator? I'm a little worried about the `madvise` traffic being increased for the pooling allocator since we already know it's a source of slow tlb-shootdowns, and I'm afraid that any cost of that will be hidden by the other performance gains in this PR so we can't accurately measure the strategy of `memset`-on-initialize vs `madvise`-to-zero.

Reply:
Modified to take a `prezeroed` flag as suggested -- good idea!

I'm a little hesitant to settle on a `memset`-only design even for the pooling allocator, as my earlier measurements with a large module (spidermonkey.wasm) showed that that became really quite hot; that's why I had gone to the initialized-bitmap instead (1 bit vs 24 bytes to zero --> 192x less). But I will benchmark this to be sure on Monday.

Alternatively, I think that if we move to a single-slot-per-instance design for the pooling allocator, such that we have one `madvise` for everything (memories, tables), stretching that range to include a few extra pages for the Instance and VMContext should be no big deal (it will share one mmap-lock acquire and one set of IPIs), so I suspect that madvise-cost may not have to be a huge consideration in the future. Of course getting there implies removing uffd first, since it depends on the separate-pools design, so we would have to make that decision first.